python/pyarrow/_fs.pyx

```diff
@@ -652,3 +636,6 @@ cdef class _MockFileSystem(FileSystem):
     cdef init(self, const shared_ptr[CFileSystem]& wrapped):
         FileSystem.init(self, wrapped)
         self.mockfs = <CMockFileSystem*> wrapped.get()
+
+    def __reduce__(self):
+        return _MockFileSystem, tuple()
```

Comment:
Side note: not sure if pickling this class can actually make sense if it doesn't preserve the data written to it (cfr https://github.com/apache/arrow/pull/6655#discussion_r394359544, creating from an URI gives a similar problem)

EDIT: I see below that you made then evaluate unequal for that reason

Reply:
Maybe it'd be better to just raise an exception if someone tries to pickle a _MockFileSystem

Reply:
Removed it, so it raises a TypeError if someone tries to pickle.